Civi/Core/Url.php

```diff
@@ -0,0 +1,690 @@
+        case 'a':
+          $this->preferFormat = 'absolute';
+          break;
+
+        // (r)elative url
+        case 'r':
+          $this->preferFormat = 'relative';
+          break;
+
+        // (h)tml encoding
+        case 'h':
+          $this->htmlEscape = TRUE;
+          break;
+
+        // (t)ext encoding (canonical URL form)
+        case 't':
+          $this->htmlEscape = FALSE;
+          break;
+
+        // (s)sl
+        case 's';
+          $this->ssl = TRUE;
+          break;
+
+        // (c)ache code for resources
+        case 'c':
+          $this->cacheCode = TRUE;
+          break;
+      }
+    }
+    return $this;
+  }
+
+  /**
+   * Render the final URL as a string.
+   *
+   * @return string
+   */
+  public function __toString(): string {
+    $userSystem = \CRM_Core_Config::singleton()->userSystem;
+    $preferFormat = $this->getPreferFormat() ?: static::detectFormat();
+    $scheme = $this->getScheme();
+
+    if ($scheme === NULL || $scheme === 'current') {
+      $scheme = static::detectScheme();
+    }
+
+    if ($scheme === 'default') {
+      $scheme = \CRM_Core_Menu::isPublicRoute($this->getPath()) ? 'frontend' : 'backend';
+    }
+
+    switch ($scheme) {
+      case 'frontend':
+      case 'service':
+        $result = $userSystem->url($this->getPath(), $this->getQuery(), $preferFormat === 'absolute', $this->composeFragment(), TRUE, FALSE, FALSE);
+        break;
+
+      case 'backend':
+        $result = $userSystem->url($this->getPath(), $this->getQuery(), $preferFormat === 'absolute', $this->composeFragment(), FALSE, TRUE, FALSE);
+        break;
+
+      case 'assetBuilder':
+        $assetName = $this->getPath();
+        $assetParams = [];
+        parse_str('' . $this->getQuery(), $assetParams);
+        $result = \Civi::service('asset_builder')->getUrl($assetName, $assetParams);
+        break;
+
+      case 'asset':
+        if (preg_match(';^\[([\w\.]+)\](.*)$;', $this->getPath(), $m)) {
+          [, $var, $rest] = $m;
+          $result = rtrim(\Civi::paths()->getVariable($var, 'url'), '/');
+          if ($preferFormat === 'relative') {
+            $result = \CRM_Utils_Url::toRelative($result);
+          }
+          $result .= $rest . $this->composeSuffix();
+        }
+        else {
+          throw new \RuntimeException("Malformed asset path: {$this->getPath()}");
+        }
+        break;
+
+      case 'ext':
+        $parts = explode('/', $this->getPath(), 2);
+        $result = \Civi::resources()->getUrl($parts[0], $parts[1] ?? NULL, FALSE) . $this->composeSuffix();
+        break;
+
+      default:
+        throw new \RuntimeException("Unknown URL scheme: {$this->getScheme()}");
+    }
+
+    if ($this->cacheCode) {
+      $result = \Civi::resources()->addCacheCode($result);
+    }
+
+    // TODO decide if the current default is good enough for future
+    $ssl = $this->getSsl() ?: \CRM_Utils_System::isSSL();
+    if ($ssl && str_starts_with($result, 'http:')) {
```

Comment:
We'll need a shim for `str_starts_with()` otherwise CiviCRM will require PHP 8.0+.

Reply:
@christianwach Yup, that's correct. [This shim](https://github.com/civicrm/civicrm-core/blob/5.63/composer.json#L100-L104) should take care of it, right?  (specifically https://github.com/symfony/polyfill-php80).

Reply:
Ah cool, missed that.